### PR TITLE
implement callback function for update

### PR DESCRIPTION
**Pull request by seidler2547**
JKBMS over Bluetooth sends updates by itself, so we don't need to poll it. This patch makes it publish new values to dbus as soon as they arrive via Bluetooth.

### CODE REVIEW

**Comment by mr-manuel** [comment]
Thanks for helping optimizing the code! I will review it in the next days.

File: etc/dbus-serialbattery/dbus-serialbattery.py [hunk 1]
```diff
@@ -152,7 +152,8 @@ def get_port() -> str:
         sys.exit(1)
 
     # Poll the battery at INTERVAL and run the main loop
```
Comment (mr-manuel) [inline etc/dbus-serialbattery/dbus-serialbattery.py:154]: Could you please also update the comment?

**Comment by mr-manuel** [comment]
@seidler2547 since this change the JKBMS_BLE driver gets stale, if it's loosing the connection. Because the loop is not executed anymore after disconnect. How can we solve this?

```bash
2023-06-15 12:41:45.074095500 INFO:SerialBattery:processing frame with battery cell info
2023-06-15 12:41:45.080366500 INFO:SerialBattery:--> DBusHelper: publish_battery()
2023-06-15 12:41:45.081041500 INFO:SerialBattery:--> JKBMS_BLE: refresh_data()
2023-06-15 12:41:45.081637500 INFO:SerialBattery:--> JKBMS_BRN: get_status()
2023-06-15 12:41:45.613783500 INFO:SerialBattery:processing frame with battery cell info
2023-06-15 12:41:45.620835500 INFO:SerialBattery:--> DBusHelper: publish_battery()
2023-06-15 12:41:45.621594500 INFO:SerialBattery:--> JKBMS_BLE: refresh_data()
2023-06-15 12:41:45.622258500 INFO:SerialBattery:--> JKBMS_BRN: get_status()
2023-06-15 12:41:46.154417500 INFO:SerialBattery:processing frame with battery cell info
2023-06-15 12:41:46.160787500 INFO:SerialBattery:--> DBusHelper: publish_battery()
2023-06-15 12:41:46.161481500 INFO:SerialBattery:--> JKBMS_BLE: refresh_data()
2023-06-15 12:41:46.162093500 INFO:SerialBattery:--> JKBMS_BRN: get_status()
2023-06-15 12:41:46.709338500 INFO:SerialBattery:processing frame with battery cell info
2023-06-15 12:41:46.720693500 INFO:SerialBattery:--> DBusHelper: publish_battery()
2023-06-15 12:41:46.721490500 INFO:SerialBattery:--> JKBMS_BLE: refresh_data()
2023-06-15 12:41:46.722210500 INFO:SerialBattery:--> JKBMS_BRN: get_status()
2023-06-15 12:41:46.897818500 INFO:SerialBattery:--> JKBMS_BRN: asy_connect_and_scrape(): Ended
2023-06-15 12:41:46.899188500 INFO:SerialBattery:--> self.main_thread.is_alive(): True
2023-06-15 12:41:47.233798500 [CHG] Device C8:47:8C:E6:E2:00 RSSI: -48
2023-06-15 12:41:47.233807500 [CHG] Device C8:47:8C:E6:E2:00 RSSI: -58
2023-06-15 12:41:47.233810500 [CHG] Device C8:47:8C:E6:E2:00 RSSI: -48
2023-06-15 12:41:47.233812500 [CHG] Device C8:47:8C:E6:E2:00 RSSI is nil
```

When `asy_connect_and_scrape()` ends, then `publish_battery()` is not executed anymore and cannot check the state of the driver anymore. Unfortunately I did not yet understand the exact reason.

**Comment by seidler2547** [comment]
Interesting. Is it because Bluetooth is restarted maybe? I thought about this situation before, but my understanding was that then the driver will exit and will automatically be restarted. Does that not happen? I am running this code and the driver has an uptime of almost 10 days now, never lost connection. But it's in the basement, with hardly any interference, not even WiFi connected.

**Comment by mr-manuel** [comment]
Yes it is, but on purpose. The driver has to exit, when no new data is received. 

I can give you access to a system with two JKBMS connected over bluetooth. Contact me on Discord for that. There has to be some mechanism that is watching, if the data is received or not else we are blind and do not know if the driver is working or not.

Best would be to implement something that executes `poll_battery()`, if it's no executed after one second.

I tried to recover in the Bluetooth threat, but without luck.
```bash
2023-06-15 13:21:25.253713500 INFO:SerialBattery:--> dbus-serialbattery: poll_battery()
2023-06-15 13:21:25.254484500 INFO:SerialBattery:--> DBusHelper: publish_battery()
2023-06-15 13:21:25.255286500 INFO:SerialBattery:--> JKBMS_BLE: refresh_data()
2023-06-15 13:21:25.259082500 INFO:SerialBattery:--> dbus-serialbattery: poll_battery() - End...
2023-06-15 13:21:25.787483500 INFO:SerialBattery:--> dbus-serialbattery: poll_battery()
2023-06-15 13:21:25.788169500 INFO:SerialBattery:--> DBusHelper: publish_battery()
2023-06-15 13:21:25.788782500 INFO:SerialBattery:--> JKBMS_BLE: refresh_data()
2023-06-15 13:21:25.792483500 INFO:SerialBattery:--> dbus-serialbattery: poll_battery() - End...
2023-06-15 13:21:25.976847500 INFO:SerialBattery:--> JKBMS_BRN: asy_connect_and_scrape(): Ended
2023-06-15 13:21:25.978959500 INFO:SerialBattery:--> self.main_thread.is_alive(): True
2023-06-15 13:21:31.337952500 INFO:SerialBattery:--> asy_connect_and_scrape(): error while connecting to bt: [org.freedesktop.DBus.Error.NoReply] Message recipient disconnected from message bus without replying
2023-06-15 13:21:31.339260500 INFO:SerialBattery:--> JKBMS_BRN: asy_connect_and_scrape(): Ended
2023-06-15 13:21:31.340377500 INFO:SerialBattery:--> self.main_thread.is_alive(): True
2023-06-15 13:21:31.348290500 INFO:SerialBattery:--> asy_connect_and_scrape(): error while connecting to bt: No Bluetooth adapters found.
2023-06-15 13:21:31.352594500 INFO:SerialBattery:--> JKBMS_BRN: asy_connect_and_scrape(): Ended
2023-06-15 13:21:31.353833500 INFO:SerialBattery:--> self.main_thread.is_alive(): True
2023-06-15 13:21:31.357257500 INFO:SerialBattery:--> asy_connect_and_scrape(): error while connecting to bt: No Bluetooth adapters found.
```

**Comment by seidler2547** [comment]
Ah, I see. I thought that it would exit, but it doesn't. The most sensible thing in this case would be to call `sys.exit(1)` here: https://github.com/Louisvdw/dbus-serialbattery/blob/6441e56ba18ec9f42a80f9218e1ae46d18b9b7d8/etc/dbus-serialbattery/bms/jkbms_brn.py#L377 because otherwise it doesn't seem to recover? I think I did some testing with not setting `self.run` to `False` here: https://github.com/Louisvdw/dbus-serialbattery/blob/6441e56ba18ec9f42a80f9218e1ae46d18b9b7d8/etc/dbus-serialbattery/bms/jkbms_brn.py#L363 but it doesn't always recover by itself. But when I restarted the driver then it was always able to reconnect.

About discord: I've been trying to create an account for years but somehow I (or my internet connection) is always so "untrustworthy" that it won't let me.

**Comment by mr-manuel** [comment]
`sys.exit(1)` will only exit the bluetooth thread and not the main thread. I already tried that. The `mainloop.run()` has to be stopped, if the bluetooth connection does not get any data.

When we are using it this way, we are missing the whole error handling part of `publish_battery()`. Maybe @idstein has another idea how we can handle that?

Providing the data with callback instead of polling seem to not let the BlueZ driver stack crash. @baranator and I though about a buffer overflow or something like that. Maybe this is prevented this way.

**Comment by seidler2547** [comment]
Yes, I also think the two changes - not scanning and not polling - are beneficial. I had a massive memory leak in dbus-daemon and it's completely gone now. I think there's a simple way to prevent the driver from stalling, I'll see if I can whip it out and submit a PR soon.

**Pull request merged**